## 七月在线机器学习实训营三期第七周(机器学习-模型调优与融合)考试
#### 考试说明:
- 起止时间：请同学在2018年3月23日至3月25日期间完成，最晚提交时间本周日（3月25日24时之前）结束，<b>逾期不接受补考,该考试分数计入平时成绩</b>
- 考试方式：请同学<font color=red><b>拷贝</b></font>该试卷至自己姓名的目录后，将文件更名为同学姓名拼音-exam7后，进行作答。例如wangwei-exam7
- 提交格式：请同学新建自己姓名全拼的文件夹，将该试卷，数据文件，zip文件等相关考试文件，放置此目录下。将该目录<b>移动</b>至/0.Teacher/Exam/7/目录下
- 注意事项：为确保同学们真正了解自身对本周课程的掌握程度，<font color=red><b>请勿翻阅，移动，更改</b></font>其它同学试卷。如发现按0分处理
- 请同学在下方同学姓名处填写自己的姓名，批改人和最终得分不用填写

- 同学姓名:<u>于洋</u>  
- 批改人： David
- 最终得分:100

<center><h1>####答卷开始####</h1></center>
<hr>

## 简答题(共6题，1，2题每题5分，后4题每题10分，共计50分)
- note： 48

#### 1. 机器学习任务中，通常会将给定数据切分为训练集，验证集，测试集，请回答这三类数据集各自的用途。

- 训练集：主要是用来拟合模型的，用来训练模型。<br>
- 验证集：作用是当通过训练集训练出多个模型后，为了能找出效果最佳的模型，使用各个模型对验证集数据进行预测，并记录模型准确率。选出效果最佳的模型所对应的参数，即用来调整模型参数。<br>
- 测试集：通过训练集和验证集得出最优模型后，使用测试集进行模型预测。用来衡量该最优模型的性能和分类能力。即可以把测试集当做从来不存在的数据集，当已经确定模型参数后，使用测试集进行模型性能评价。<br>

#### 2.如何将数据转换成xgboost内置用法的libsvm数据格式，以及该格式数据是如何解读的?

答：可以使用xgboost的DMatrix方法将数据转换成xgboost内置用法的libsvm数据格式。libsvm数据格式是Label 1:value 2:value …形式的，这里的Label是类别的标识，value是数据（特征值），1：表示的这个值（特征）出现的位置，相当于索引，这样对于特征值为0的位置就可以不进行表示，例如：1 1:3.14 3:2.15  这里索引2的位置的特征值为0，所以不需要表示，这样也节省了内存的使用。

#### 3. 对于自动特征选择，通常有哪三类方法？试写出每类方式的思想，以及在sklearn中的代码实现（面试题）

- note :代码实现呢，但最后实验题那么好，就不忍心在这个细节 上扣了，直接满分。

答：特征选择的三类方法：
- 1.过滤型：过滤式方法先对数据集进行特征选择，再训练学习器。使用sklearn.SelectBest(k个)选择k个特征，使用SelectPercentile(百分之多少)选择百分之多少个数据特征。
- 2.包裹型：包裹式特征选择直接把最终要使用的学习器性能作为特征子集的评价标准。sklearn.feature_selection.RFE()进行特征子集的筛选。
- 3.嵌入型：嵌入式特征选择将特征选择过程和机器训练过程融合为一体。两者在同一优化过程中完成，即在学习器训练过程中自动进行了特征选择。sklearn.feature.selection.SelectFromModel进行根据模型分析特征的重要性

#### 4.网格搜索交叉验证的作用是什么，并简述网格搜索交叉验证是如何运行的？ 

答：
- 网格搜索是指定参数值的一种穷举搜索方法，通过将估计函数的参数通过交叉验证的方法进行优化来得到最优的学习算法。即：将各个参数可能的取值进行排列组合，列出所有可能的组合结果生成“网格”。<br>
- 交叉验证是先将数据集D划分为k个大小相似的互斥子集，即D = D1 ∪ D2 ∪...∪ Dk，而且要满足Di ∩ Dj = ∅（i ≠ j），每个子集都通过分层采样得到，然后每次使用k - 1个子集的并集作为训练集，余下的那个子集作为测试集，这样进行k次训练要和测试，返回k个测试结果的均值作为模型评估的结果。这种方法也叫作k折交叉验证法<br>

#### 5.工业界上模型整合有三大类方式？试简述每类方式其思想？（面试题）

答：工业界模型整合（融合）分为Boosting、Bagging、Stacking。
- Boosting
Boosting方法是一种用来提高弱分类算法准确度的方法,这种方法通过构造一个预测函数系列,然后以一定的方式将他们组合成一个预测函数。代表算法为Adaboost。
- Bagging
Bagging选择样本时是随机的，在选择划分属性时还是每一个属性都会进行考量，选取最优的属性进行划分。其思想为：1.给定一个弱学习算法,和一个训练集;2.单个弱学习算法准确率不高;3.将该学习算法使用多次,得出预测函数序列,进行投票;4.最后结果准确率将得到提高.
- Stacking  
将训练好的所有基模型对整个训练集进行预测，第j个基模型对第i个训练样本的预测值将作为新的训练集中第i个样本的第j个特征值，最后基于新的训练集进行训练。同理，预测的过程也要先经过所有基模型的预测形成新的测试集，最后再对测试集进行预测。

#### 6.  我们可以将xgboost的中众多参数分类为哪三类？请写出哪些参数可以用什么方式用来控制过拟合？

答：xgboost的中众多参数分类通用参数、booster参数、任务型参数 
- eta：学习率， 为了防止过拟合，更新过程中用到的收缩步长。在每次提升计算之后，算法会直接获得更新的权重。通过减少特征的权重使提升计算的过程更加保守。  
- max_depth：用于设置树的最大树深，max_depth越大，此值设置可以增加生成树的深度，随着深度增大学到的特征越详细，但是不能设置过大，过大会造成模型将样本的不相关属性值都学到，例如日期。 
- subsample：表示观测的子样本的比率，将其设置为0.5意味着xgboost将随机抽取一半观测用于树的生长，这将有助于防止过拟合现象。

## 实验题(共2题，每题25分，共计50分)

### 1. 使用XGBoost内置方式，导入iris数据完成分类问题（要求以不同参数设置xgboost运行并比对），最后给出实验总结

In [26]:
# 导入所需的库
import xgboost as xgb
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, mean_squared_error
from sklearn.metrics import accuracy_score
# 加载数据
iris = load_iris()

In [4]:
X = iris.data
Y = iris.target

In [6]:
# 分割数据集
X_train, X_test, y_train, y_test = train_test_split(X, Y, random_state=0)

In [7]:
d_train = xgb.DMatrix(data=X_train, label=y_train)
d_test = xgb.DMatrix(data=X_test, label=y_test)

In [13]:
#设置参数
param = {'max_depth':3, 'eta':0.05, 'objective':'multi:softmax', 'num_class':3}
#第一次训练
xgbst = xgb.cv(param, d_train, nfold=5,seed = 2018)
xgbst

test-merror-mean  test-merror-std  train-merror-mean  train-merror-std
0          0.044664         0.000969           0.015630          0.005485
1          0.044664         0.000969           0.013408          0.008362
2          0.035969         0.018001           0.013408          0.008362
3          0.035969         0.018001           0.013408          0.008362
4          0.035969         0.018001           0.013408          0.008362
5          0.035969         0.018001           0.011161          0.009994
6          0.035969         0.018001           0.011161          0.009994
7          0.035969         0.018001           0.013383          0.008369
8          0.035969         0.018001           0.013383          0.008369
9          0.035969         0.018001           0.013383          0.008369

使用初始参数进行评估：最大树深：3，学习率：0.05，随着迭代的次数，验证集平均错误率从0.44664降到0.35969，训练集平均错误率从0.15630降到0.11161，又提升到0.13383，随着迭代轮数先降低后又反弹。

In [ ]:
#在第一次的基础上改变学习率到0.1，其他参数不变

In [18]:
param = {'max_depth':3, 'eta':0.5, 'objective':'multi:softmax', 'num_class':3}
#第二次训练
xgbst = xgb.cv(param, d_train, nfold=5,seed = 2018)
xgbst

test-merror-mean  test-merror-std  train-merror-mean  train-merror-std
0          0.044664         0.000969           0.015630          0.005485
1          0.044664         0.000969           0.013383          0.008369
2          0.035969         0.018001           0.013408          0.008362
3          0.035969         0.018001           0.013408          0.008362
4          0.045059         0.028759           0.011161          0.007027
5          0.035969         0.018001           0.008939          0.004470
6          0.035969         0.018001           0.004444          0.005443
7          0.035969         0.018001           0.002222          0.004444
8          0.035969         0.018001           0.000000          0.000000
9          0.035969         0.018001           0.000000          0.000000

在第一次基础上使用参数进行评估：最大树深：3，学习率：0.5，随着迭代的次数，验证集平均错误率从0.44664降到0.35969，中间又一次反弹但最终结果还是降低的；训练集平均错误率从0.15630降到0，且中间没有出现反弹，训练标准差也降到0，说明改变学习率之后训练集平均错误率有所提升。

In [19]:
#在第二次的基础上改变最大树深为5，其他参数不变
param = {'max_depth':5, 'eta':0.5, 'objective':'multi:softmax', 'num_class':3}
#第三次训练
xgbst = xgb.cv(param, d_train, nfold=5,seed = 2018)
xgbst

test-merror-mean  test-merror-std  train-merror-mean  train-merror-std
0          0.044664         0.000969           0.015630          0.005485
1          0.044664         0.000969           0.013383          0.008369
2          0.035969         0.018001           0.013408          0.008362
3          0.035969         0.018001           0.011161          0.007027
4          0.035969         0.018001           0.011161          0.007027
5          0.035969         0.018001           0.006692          0.005464
6          0.035969         0.018001           0.004444          0.005443
7          0.035969         0.018001           0.002222          0.004444
8          0.035969         0.018001           0.000000          0.000000
9          0.035969         0.018001           0.000000          0.000000

在第二次基础上使用参数进行评估：最大树深：5，学习率：0.5，随着迭代的次数，验证集平均错误率从0.44664降到0.35969，；训练集平均错误率从0.15630降到0，且中间没有出现反弹，较上一次比较，模型的平均错误率和收敛速度没有得到提升，说明改变树深并没有带来模型性能的提升。

In [21]:
#在第三次的基础上改变最大树深为2，其他参数不变
param = {'max_depth':2, 'eta':0.5, 'objective':'multi:softmax', 'num_class':3}
#第四次训练
xgbst = xgb.cv(param, d_train, nfold=5,seed = 2018)
xgbst

test-merror-mean  test-merror-std  train-merror-mean  train-merror-std
0          0.062846         0.022925           0.035730          0.008404
1          0.045059         0.028759           0.022322          0.007107
2          0.045059         0.028759           0.022322          0.014095
3          0.045059         0.028759           0.024569          0.004570
4          0.035969         0.018001           0.013433          0.013069
5          0.035969         0.018001           0.008964          0.008402
6          0.035969         0.018001           0.006742          0.008989
7          0.035969         0.018001           0.004494          0.005504
8          0.035969         0.018001           0.000000          0.000000
9          0.035969         0.018001           0.000000          0.000000

在第三次基础上使用参数进行评估：最大树深：2，学习率：0.5，随着迭代的次数，验证集平均错误率从0.62846降到0.35969，；训练集平均错误率从0.35730降到0，较上一次比较，模型的平均错误率和收敛速度没有得到提升，说明改变树深并没有带来模型性能的提升。所以最终确定参数选择为最大树深：3，学习率：0.5

In [28]:
#对测试集数据进行验证
final = xgb.train(param, d_train)
d_test = xgb.DMatrix(data=X_test, label=y_test)
y_predict = final.predict(d_test)
print('Accuracy:',accuracy_score(y_test, y_predict))

Accuracy: 0.9736842105263158


最终预测结果的准确率为97.368%。说明模型的学习效果比较好。

### 2. 使用XGBoost的sklearn接口，对KaggleCredit2数据完成信用卡欺诈项目的建模及分析（要求以不同参数设置xgboost运行并比对），最后给出实验总结

- KaggleCredit2数据文件 位于/home/sxy-s3/0.Teacher/Exam/KaggleCredit2.csv.zip，请勿复制或移动该文件

In [35]:
#倒入库
import pandas as pd
import xgboost as xgb
from sklearn.model_selection import GridSearchCV,train_test_split
from sklearn.metrics import confusion_matrix,accuracy_score
import warnings
import zipfile
warnings.filterwarnings('ignore')

In [36]:
with zipfile.ZipFile('/home/sxy-s3/0.Teacher/Exam/KaggleCredit2.csv.zip', 'r') as z:
    f = z.open('KaggleCredit2.csv')
data = pd.read_csv(f, index_col=0)
data.head()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines   age  \
0                 1                              0.766127  45.0   
1                 0                              0.957151  40.0   
2                 0                              0.658180  38.0   
3                 0                              0.233810  30.0   
4                 0                              0.907239  49.0   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                   2.0   0.802982         9120.0   
1                                   0.0   0.121876         2600.0   
2                                   1.0   0.085113         3042.0   
3                                   0.0   0.036050         3300.0   
4                                   1.0   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                             13.0                      0.0   
1                              4.0                      0.0   
2                              2.0                      1.0   
3                              5.0                      0.0   
4                              7.0                      0.0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                           6.0                                   0.0   
1                           0.0                                   0.0   
2                           0.0                                   0.0   
3                           0.0                                   0.0   
4                           1.0                                   0.0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [40]:
#查看是否有缺失值
data.isnull().sum()

SeriousDlqin2yrs                           0
RevolvingUtilizationOfUnsecuredLines       0
age                                     4267
NumberOfTime30-59DaysPastDueNotWorse       0
DebtRatio                                  0
MonthlyIncome                              0
NumberOfOpenCreditLinesAndLoans            0
NumberOfTimes90DaysLate                    0
NumberRealEstateLoansOrLines               0
NumberOfTime60-89DaysPastDueNotWorse       0
NumberOfDependents                      4267
dtype: int64

In [43]:
#使用平均年龄填充年龄缺失值
data['age'] = data['age'].fillna(data['age'].mean())
# 使用sklearn填充NumberOfDependents的缺失值
# 最高频率出现的填充空值
from sklearn.preprocessing import Imputer
imp = Imputer(strategy='most_frequent',axis=0)
data['NumberOfDependents'] = imp.fit_transform(data[['NumberOfDependents']])

In [44]:
data.isnull().sum()

SeriousDlqin2yrs                        0
RevolvingUtilizationOfUnsecuredLines    0
age                                     0
NumberOfTime30-59DaysPastDueNotWorse    0
DebtRatio                               0
MonthlyIncome                           0
NumberOfOpenCreditLinesAndLoans         0
NumberOfTimes90DaysLate                 0
NumberRealEstateLoansOrLines            0
NumberOfTime60-89DaysPastDueNotWorse    0
NumberOfDependents                      0
dtype: int64

In [45]:
data.dropna(inplace=True)
y = data['SeriousDlqin2yrs']
X = data.drop('SeriousDlqin2yrs', axis=1)
X_train, X_test, y_train, y_test = train_test_split(X, y,\
                                                    test_size = 0.3,\
                                                    random_state = 0)

In [46]:
param_grid = {'max_depth': [2,4],
              'n_estimators': [50,100],
               'learning_rate': [0.01,0.1],
               'subsample': [0.8, 0.9],
               'colsample_bylevel':[0.8, 0.9]
             }
xgb_model = xgb.XGBClassifier()
clf = GridSearchCV(xgb_model, param_grid)

clf.fit(X_train,y_train)

GridSearchCV(cv=None, error_score='raise',
       estimator=XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=1),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'n_estimators': [50, 100], 'subsample': [0.8, 0.9], 'max_depth': [2, 4], 'learning_rate': [0.01, 0.1], 'colsample_bylevel': [0.8, 0.9]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [47]:
print(clf.best_score_)
print(clf.best_params_)

0.9367535425101214
{'n_estimators': 50, 'subsample': 0.9, 'colsample_bylevel': 0.9, 'learning_rate': 0.1, 'max_depth': 4}


通过交叉验证我们可以看到模型迭代轮次为50轮，观测的子样本的比率为0.9，对列数的采样的占比为0.9，学习率为0.1，最大树深为4，这是饿哦们通过交叉验证得到的最优参数和最好结果0.9367535425101214

In [49]:
#进行预测
y_pred = clf.best_estimator_.predict(X_test)

In [50]:
print(accuracy_score(y_test, y_pred))

0.9349077490774907


可以看到最终的预测结果和真实结果准确率达到了93.491%，和用网格搜索交叉验证得到的最优结果相差无几，但是还可以再进行优化。使用xgboost提供的sklearn接口和交叉验证很方便的得到训练模型的一些最优参数和结果，省去了我们手动调参的时间，在这里我们没有对样本的特征做过多的选择与分析，在实际工作共特征工程是很关键的一步，我们还需要对特征进行分析和筛选，已达到模型更优的目的。

### 本阶段课程意见反馈题(非必答，不送分)
#### 请同学围绕以下两点进行回答：
- 自身总结：请您对您自己在本周课程的学习，收获，技能掌握等方面进行一次总结 ，也包括有哪些方面存在哪些不足，欠缺，困惑。作为将来回顾学习路径时的依据。


- 课程反馈：请就知识点，进度，难易度，教学方式，考试方式及难易度等方面向我们反馈，督促我们进行更有效的改进。

<hr>
<center><h1>####答卷结束####</h1></center>